### PR TITLE
Annotated Type Loader

### PR DESCRIPTION
**Pull request by jyluo**

### CODE REVIEW

File: checker/examples/units-extension/Demo.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import org.checkerframework.checker.units.UnitsTools;
 import org.checkerframework.checker.units.qual.*;
+import qual.*;
```
Comment (wmdietl) [inline checker/examples/units-extension/Demo.java]: Can you replace both import \* with specific imports?

File: checker/examples/Makefile [hunk 2]
```diff
@@ -4,7 +4,6 @@ all: compile
 
 compile:
 	$(JAVAC) *.java
-	$(JAVAC) units-extension/*.java
```
Comment (wmdietl) [inline checker/examples/Makefile:7]: Why is this no longer needed? Doesn't this at least check that the code compiles?

File: checker/examples/units-extension/Makefile [hunk 3]
```diff
@@ -1,24 +1,39 @@
...
+  qual/kHz.java
 
 JAVAOPTS=-source 7 -target 7
 
 JSR308JAVAC?=../../bin/javac
 
-all: framework-test clean
+# gets the full path to the directory of the make file, which is also the root dir of the qual folder
+# for custom projects, it is best to encode the full root path as a variable
+QUALDIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
```
Comment (wmdietl) [inline checker/examples/units-extension/Makefile]: This is the directory that contains the "qual" subdirectory, right?
QUALDIR sounds more like it is the "qual" directory.

File: checker/manual/creating-a-checker.tex [hunk 4]
```diff
@@ -194,27 +193,26 @@ \subsection{Defining the type qualifiers\label{define-type-qualifiers}}
...
 \noindent
-Write the \refqualclass{framework/qual}{TypeQualifier} meta-annotation on the annotation definition
-to indicate that the annotation represents a type qualifier
-and should be processed by the checker.
-Also write a \sunjavadocanno{java/lang/annotation/Target.html}{Target}
-meta-annotation to indicate where the annotation may be written.
+
+Write a \sunjavadocanno{java/lang/annotation/Target.html}{Target}
+meta-annotation on the annotation definition to indicate where the annotation may be written.
+All type qualifiers that users can use should have the value \<ElementType.TYPE\_USE>.
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: Should also list ElementType.TYPE_PARAMETER.

File: checker/manual/creating-a-checker.tex [hunk 5]
```diff
@@ -430,6 +424,7 @@ \subsection{Completeness of the type hierarchy\label{bottom-and-top-qualifier}}
   @interface IGJBottom { }
 \end{Verbatim}
 
+This qualifier is added to the IGJ checker manually in IGJAnnotatedTypeFactory.
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: Capitalize "checker".
Fix reference to IGJAnnotatedTypeFactory.

File: checker/manual/creating-a-checker.tex [hunk 6]
```diff
@@ -1094,31 +1087,81 @@ \section{The checker class:  Compiler interface\label{writing-compiler-interface
...
-polymorphic qualifiers), either via a
-\refqualclass{framework/qual}{TypeQualifiers} annotation or by overriding
+polymorphic qualifiers), either by placing the annotations inside a subfolder called \<qual> located in the same directory as the checker, or by overriding
 the
 \refmethodterse{framework/type}{AnnotatedTypeFactory}{createSupportedTypeQualifiers}{--}
-method.  Each argument to \<@TypeQualifiers> or value returned by
-\<createSupportedTypeQualifiers> is a class literal for a type qualifier
-whose definition bears the
-\refqualclass{framework/qual}{TypeQualifier} meta-annotation.
+method. By default, the Polymorphic Qualifier \refqualclass{framework/qual}{PolyAll} is not utilized, but can be optionally added to a checker.
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: Lower case polymorphic qualifier.

Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: Is there any harm in always adding PolyAll?
It's much more overhead to add a qualifier without TypeQualifiers annotations. So automatically adding PolyAll seems like a good change.

File: checker/manual/creating-a-checker.tex [hunk 7]
```diff
@@ -1094,31 +1087,81 @@ \section{The checker class:  Compiler interface\label{writing-compiler-interface
...
+
+There's four recommended ways to write an override implementation of \refmethodterse{framework/type}{AnnotatedTypeFactory}{createSupportedTypeQualifiers}{--}
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: There are four.

File: checker/manual/creating-a-checker.tex [hunk 8]
```diff
@@ -1094,31 +1087,81 @@ \section{The checker class:  Compiler interface\label{writing-compiler-interface
...
+        new HashSet<Class<? extends Annotation>>(
+            Arrays.asList(H1Top.class, H1S1.class, H1S2.class,
+                H1Bot.class, H2Top.class, H2S1.class, H2S2.class,
+                H2Bot.class, H1Poly.class, H2Poly.class)));
+}
+\end{Verbatim}
+
+The use of a manual list is necessary for any qualifiers that are not located in
+the \<qual> folder, or for qualifiers which do not or cannot have the
+\<@Target(\{ElementType.TYPE\_USE\})> meta-annotation (eg some Bottoms).
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: "eg some Bottoms" -> "e.g. qualifiers that should not be directly used in source code, but are still part of the hierarchy"

File: checker/manual/creating-a-checker.tex [hunk 9]
```diff
@@ -1094,31 +1087,81 @@ \section{The checker class:  Compiler interface\label{writing-compiler-interface
...
+    return Collections.unmodifiableSet(
+        new HashSet<Class<? extends Annotation>>(
+            Arrays.asList(A.class, B.class, C.class, D.class, E.class, F.class)));
+}
+\end{Verbatim}
+
+\end{enumerate}
+
+Each value returned by \<createSupportedTypeQualifiers> is a class literal for a type qualifier
+whose definition bears the \<@Target(\{ElementType.TYPE\_USE\})> meta-annotation.
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: Not all have TYPE_USE, as mentioned in one of the previous points.

File: checker/manual/creating-a-checker.tex [hunk 10]
```diff
@@ -1094,31 +1087,81 @@ \section{The checker class:  Compiler interface\label{writing-compiler-interface
...
+protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+    return Collections.unmodifiableSet(
+        new HashSet<Class<? extends Annotation>>(
+            Arrays.asList(A.class, B.class, C.class, D.class, E.class, F.class)));
+}
+\end{Verbatim}
+
+\end{enumerate}
+
+Each value returned by \<createSupportedTypeQualifiers> is a class literal for a type qualifier
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: I would only call "X.class" a class literal. You mean "<Class> instance"?

File: checker/manual/manual.bbl [hunk 11]
```diff
@@ -45,14 +45,13 @@ Werner Dietl, Michael~D. Ernst, and Peter M{\"u}ller.
   Conference}, pages 333--357, Lancaster, UK, July~27--29, 2011.
 
 \bibitem[EJM{\etalchar{+}}14]{ErnstJMDPRKBBHVW2014}
-Michael~D. Ernst, Ren{\'e} Just, Suzanne Millstein, Werner Dietl, Stuart
+Michael~D. Ernst, Ren{\'e} Just, Suzanne Millstein, Werner~M. Dietl, Stuart
```
Comment (wmdietl) [inline checker/manual/manual.bbl]: Why the update to this file??
I don't have the "M." in papers.
You loose page numbers in a paper below and completely remove one paper.
Undo this file change.

File: checker/manual/troubleshooting.tex [hunk 12]
```diff
@@ -741,9 +736,32 @@ \subsection{Build the Checker Framework\label{building}}
 \end{Verbatim}
 % $ to unconfuse Emacs LaTeX mode
 
+Once \<checker.jar> is built. You may wish to put the Checker Framework \<javac> even earlier on
```
Comment (wmdietl) [inline checker/manual/troubleshooting.tex]: Strange first - second sentence?

File: checker/manual/troubleshooting.tex [hunk 13]
```diff
@@ -683,11 +683,6 @@ \subsection{Build the Type Annotations compiler\label{building-compiler}}
   export PATH=$JSR308/jsr308-langtools/dist/bin:${PATH}
 \end{Verbatim}
 
-You may wish to later put the Checker Framework \<javac> even earlier on
```
Comment (wmdietl) [inline checker/manual/troubleshooting.tex]: Are these changes by you, Jeff? Or is this from some other changeset that I didn't have?
This feels like a completely independent change to the manual. Can you put it in a separate pull request?

File: checker/manual/troubleshooting.tex [hunk 14]
```diff
@@ -741,9 +736,32 @@ \subsection{Build the Checker Framework\label{building}}
...
+your path. The Checker Framework's \code{javac} ensures that \<checker.jar> 
+is on your classpath and bootclasspath, but is otherwise
+identical to the Type Annotations compiler:
+
+\begin{Verbatim}
+export PATH=$JSR308/checker-framework/checker/bin:$JSR308/jsr308-langtools/dist/bin:${PATH}
+\end{Verbatim}
+
+If you are developing Checkers and wish to use the latest builds of your source files compiled by
+eclipse, you can use the development version of javac instead. This version will use compiled
```
Comment (wmdietl) [inline checker/manual/troubleshooting.tex]: Upper case Eclipse.

File: checker/manual/troubleshooting.tex [hunk 15]
```diff
@@ -741,9 +736,32 @@ \subsection{Build the Checker Framework\label{building}}
...
+classes from \<dataflow/build>, \<javacutil/build>, \<framework/build>, and \<checker/build>
+in the \<checker-framework> directory instead of checker.jar, and by default will print all stack
```
Comment (wmdietl) [inline checker/manual/troubleshooting.tex]: "print all stack errors" -> "print stack traces for all errors"

File: checker/manual/units-checker.tex [hunk 16]
```diff
@@ -241,8 +252,21 @@ \section{Running the Units Checker\label{units-running}}
...
+You can also provide a set of folders that contain units qualifiers using the
+\code{-AunitsDirs} option, and also adding the folders to the bootclass path, for example:
+
+\begin{alltt}
+  javac -Xbootclasspath/p:\textit{/full/path/to/myproject:/full/path/to/otherLib} \ttbs
+        -processor org.checkerframework.checker.units.UnitsChecker \ttbs
+        \textit{-AunitsDirs=/full/path/to/myproject:/full/path/to/otherLib} MyFile.java ...
 \end{alltt}
+
+Note that in this example, qual.MyUnit and qual.MyOtherUnit would exist in either the myproject folder or the otherLib folder.
```
Comment (wmdietl) [inline checker/manual/units-checker.tex]: Use `\<code>`

File: checker/src/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java [hunk 17]
```diff
@@ -157,6 +164,26 @@ public NullnessAnnotatedTypeFactory(BaseTypeChecker checker, boolean useFbc) {
     }
 
     @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java:169]: AbstractNullnessChecker takes a boolean whether to use raw/fbc.
Would it be more general to pass along that boolean to the Nullness ATF and use it here?

File: checker/src/org/checkerframework/checker/propkey/PropertyKeyAnnotatedTypeFactory.java [hunk 18]
```diff
@@ -52,6 +54,15 @@ public PropertyKeyAnnotatedTypeFactory(BaseTypeChecker checker) {
     }
 
     @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        // TODO: it is possible to load @PropertyKey and @UnknownPropertyKey using reflection, but it would also load @PropertyKeyBottom
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/propkey/PropertyKeyAnnotatedTypeFactory.java]: Why is this checker using Bottom instead of PropertyKeyBottom?
The shared Bottom should only be used in the Subtyping Checker.
Why have the dedicated PropertyKeyBottom and then not use it?

File: framework/src/org/checkerframework/common/subtyping/SubtypingChecker.java [hunk 19]
```diff
@@ -28,7 +28,7 @@
  *
  * @checker_framework.manual #subtyping-checker Subtying Checker
  */
-@SupportedOptions( { "quals" })
+@SupportedOptions( { "quals", "qualFolders" })
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/subtyping/SubtypingChecker.java]: This one uses "Folders" another one uses "Dirs". Do we have a precedent for either one? Consistency would be nice.

File: framework/src/org/checkerframework/common/util/report/ReportAnnotatedTypeFactory.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
+package org.checkerframework.common.util.report;
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/util/report/ReportAnnotatedTypeFactory.java]: Wouldn't it be much easier to introduce a "ReportUnqualified" annotation?

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java [hunk 21]
```diff
@@ -0,0 +1,324 @@
+package org.checkerframework.framework.type;
+
+import java.io.File;
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: Use the Eclipse import ordering from checker-framework/checker/eclipse/checker-framework.importorder

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.framework.qual.PolyAll;
+import org.checkerframework.framework.util.AnnotationBuilder;
+
+/*>>>
+import org.checkerframework.checker.nullness.qual.Nullable;
+ */
+
+public class AnnotatedTypeLoader {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: New public class needs good JavaDoc!

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java [hunk 23]
```diff
@@ -0,0 +1,324 @@
...
+    // used to build the annotation mirror from the loaded Class object
+    protected ProcessingEnvironment processingEnv;
+
+    // Stores the Resource URL of the qual directory of a Checker class, or AnnotatedTypeFactory class of a particular checker
+    URL resourceURL;
+
+    // Stores a mapping of the Annotation object and the converted AnnotationMirror
+    private Map<Class<? extends Annotation>, AnnotationMirror> loadedAnnotations;
+
+    // constructor for loading type qualifiers defined for a checker
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: Use JavaDoc, not code comments.

File: framework/src/org/checkerframework/framework/util/AnnotatedTypes.java [hunk 24]
```diff
@@ -1405,6 +1405,7 @@ public static boolean isTypeAnnotation(AnnotationMirror anno) {
         return result;
     }
 
+    @SuppressWarnings("deprecation")
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/AnnotatedTypes.java]: Is this method also deprecated? Can you already remove it?

File: framework/src/org/checkerframework/framework/util/PurityAnnotatedTypeFactory.java [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+    public PurityAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        this.postInit();
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return Collections.unmodifiableSet(
+                new HashSet<Class<? extends Annotation>>(
+                        Arrays.asList(Unqualified.class)));
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/PurityAnnotatedTypeFactory.java]: Would it be easier to use a PurityUnqualified instead?

File: checker/examples/Makefile (hunk 2 above)
Comment (jyluo) [inline checker/examples/Makefile:7]: The compilation done here would require a class path to the examples directory for the external units qualifiers to compile. Instead, in the make file within units-extension, it does a fresh compilation and runs the tests. That make file is invoked with ant target units-extension-tests.

File: framework/src/org/checkerframework/common/util/report/ReportAnnotatedTypeFactory.java (hunk 20 above)
Comment (jyluo) [inline framework/src/org/checkerframework/common/util/report/ReportAnnotatedTypeFactory.java]: I tried creating a clone of the Unqualified as ReportUnqualified. Using this clone, there's errors when running report-tests.

We can investigate why in the future. I'll keep it as Unqualified for now.

File: framework/src/org/checkerframework/framework/util/PurityAnnotatedTypeFactory.java (hunk 25 above)
Comment (jyluo) [inline framework/src/org/checkerframework/framework/util/PurityAnnotatedTypeFactory.java]: This class, and framework/util/PurityChecker doesn't seem to be used at all. There's a separate PurityChecker in dataflow/util which is used by BaseTypeVisitor.

Should I remove this class and framework/util/PurityChecker?

File: checker/examples/units-extension/Makefile [hunk 26]
```diff
@@ -1,24 +1,39 @@
...
+PROJECTDIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
+
+all: compile-for-test named-quals-test qual-folder-test clean
 
 demo:
-	$(JSR308JAVAC) $(JAVAOPTS) $(FILES)
-	@echo "***** This command is expected to produce errors on line 10 & 61:"
-	$(JSR308JAVAC) -Xbootclasspath/p:. -AprintErrorStack -processor org.checkerframework.checker.units.UnitsChecker -Aunits=Hz,kHz,Frequency Demo.java
+	$(JAVAC) $(JAVAOPTS) $(FILES)
+	@echo "***** This command is expected to produce errors on line 11 & 62:"
```
Comment (wmdietl) [inline checker/examples/units-extension/Makefile]: Should this be 16 and 66?

Comment (jyluo) [inline checker/examples/units-extension/Makefile]: Yes indeed!

File: checker/manual/creating-a-checker.tex [hunk 27]
```diff
@@ -1094,34 +1092,136 @@ \section{The checker class:  Compiler interface\label{writing-compiler-interface
...
+Override \<createSupportedTypeQualifiers()> and return an immutable set of the
+supported annotations, plus \<PolyAll>. Code example:
+\begin{Verbatim}
+@Override
+protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+return Collections.unmodifiableSet(
+    new HashSet<Class<? extends Annotation>>(
+        Arrays.asList(
+            A.class, B.class,
+            PolyAll.class)));
```
Comment (wmdietl) [inline checker/manual/creating-a-checker.tex]: Is this really worth separating into two possibilities? You have to list PolyAll explicitly, so there is no different method that you call to automatically support it.

File: checker/manual/fenum-checker.tex [hunk 28]
```diff
@@ -64,15 +64,13 @@ \section{Fake enum annotations\label{fenum-annotations}}
 file \code{\emph{MyFenum}.java}:
 
 \begin{alltt}
-package \textit{myproject}.qual;
```
Comment (wmdietl) [inline checker/manual/fenum-checker.tex:67]: Why remove the myproject package?

File: checker/manual/fenum-checker.tex [hunk 29]
```diff
@@ -120,9 +118,24 @@ \section{Running the Fenum Checker\label{fenum-running}}
...
-        \textit{-Aquals=myproject.qual.MyFenum} MyFile.java ...
+        -Aquals=\textit{qual.MyFenum} MyFile.java ...
 \end{alltt}
 
+This option specifies a list of the fully-qualified class names of annotations,
+using a comma-no-space-separated notation.
+
+You can also provide the fully-qualified paths to a set of directories
+that contain the qualifiers using the \code{-AqualDirs} option, and add
+the directories to the boot classpath, for example:
```
Comment (wmdietl) [inline checker/manual/fenum-checker.tex]: Why would I want to add the directories to the boot classpath?
Why is this not necessary for -Aquals?

File: checker/manual/subtyping-checker.tex [hunk 30]
```diff
@@ -42,26 +42,37 @@ \section{Using the Subtyping Checker\label{subtyping-using}}
...
 The annotations listed in \code{-Aquals} must be accessible to
 the compiler during compilation in the classpath.  In other words, they must
 already be compiled (and, typically, be on the javac bootclasspath)
 before you run the Subtyping Checker with \code{javac}.  It
 is not sufficient to supply their source files on the command line.
 
+\item
+\code{-AqualDirs}: this option specifies a colon-no-space-separated list of
+the fully-qualified paths to a set of directories that contain the annotations
+in the custom type system. The directories must also be added to the boot
```
Comment (wmdietl) [inline checker/manual/subtyping-checker.tex]: Here the previous paragraph mentions that the same is true for -Aquals.
Can you maybe just unify that aspect and mention it once?

File: checker/manual/creating-a-checker.tex (hunk 27 above)
Comment (jyluo) [inline checker/manual/creating-a-checker.tex]: I'll eliminate the redundant example.

File: checker/manual/fenum-checker.tex (hunk 28 above)
Comment (jyluo) [inline checker/manual/fenum-checker.tex:67]: The passages later asks them to pass the full path to myproject to -AqualDirs, thus the class loader would start looking for classes within this folder. If the qualifier is in package myproject.qual, then the folder that needs to be passed to -AqualDirs is the parent folder of myproject.

Either naming approach is fine, what would be preferred?

File: checker/manual/fenum-checker.tex (hunk 29 above)
Comment (jyluo) [inline checker/manual/fenum-checker.tex]: The directories are needed for both options. Will add the note to -Aquals as well.

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeLoader.java [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+
+    public UnitsAnnotatedTypeLoader(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    /**
+     * Custom filter for units annotations:
+     *
+     * This filter will ignore (by returning false) any units annotation which is an alias of another
+     * base unit annotation (identified via {@link UnitsMultiple} meta-annotation).
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeLoader.java]: I don't understand this logic: the qualifier is still a supported annotation class, even if it is an alias for a base unit.
If the loader doesn't load this class, can it still be written in the source code?

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeLoader.java [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+        AnnotationBuilder builder = new AnnotationBuilder(processingEnv, annoClass);
+        AnnotationMirror initialResult = builder.build();
+
+        // further refine to see if the annotation is an alias of some other SI Unit annotation
+        for (AnnotationMirror metaAnno : initialResult.getAnnotationType().asElement().getAnnotationMirrors() ) {
+            // TODO : special treatment of invisible qualifiers?
+
+            // if the annotation is a SI prefix multiple of some base unit, then return false
+            // classic Units checker does not need to load the annotations of SI prefix multiples of base units
+            if ( metaAnno.getAnnotationType().toString().equals(UnitsMultiple.class.getCanonicalName())) {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeLoader.java]: Why the String comparison instead of AnnotationUtils.areSame?

File: framework/src/org/checkerframework/common/util/report/ReportAnnotatedTypeFactory.java (hunk 20 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/common/util/report/ReportAnnotatedTypeFactory.java]: Did you set that ReportUnqualified as the default? What kind of errors did come up?
Can you create an Issue for this, so that we don't forget.

File: framework/src/org/checkerframework/framework/qual/TypeQualifiers.java [hunk 33]
```diff
@@ -6,68 +6,148 @@
...
-import javax.annotation.processing.SupportedAnnotationTypes;
-
 import org.checkerframework.framework.type.AnnotatedTypeFactory;
-import org.checkerframework.framework.type.AnnotatedTypeMirror;
 import org.checkerframework.qualframework.base.Checker;
 
 /**
- * An annotation that lists the type qualifiers supported by the annotated
- * {@code Checker}.
+ * This meta-annotation is deprecated.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/qual/TypeQualifiers.java:14]: These examples are great. But instead of putting them on the deprecated annotations, should you move them to the createSupportedTypeQualifiers method?

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java [hunk 34]
```diff
@@ -0,0 +1,408 @@
...
+        loadedAnnotations = new HashSet<Class<? extends Annotation>>();
+    }
+
+    /**
+     * Gets the set of the loaded annotation classes.
+     *
+     * @return the set of loaded annotation classes
+     */
+    public final Set<Class<? extends Annotation>> getLoadedAnnotationClasses() {
+        return loadedAnnotations;
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: Should this raise an error if classesLoaded == false?
Why doesn't the constructor call loadBundledAnnotationClasses and make that set final and immutable? Might that be easier than keeping the boolean? Are there any modifications between construction and the call of loadBundledAnnotationClasses that warrant the boolean?

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java [hunk 35]
```diff
@@ -0,0 +1,408 @@
...
+                    // class
+                    Class<? extends Annotation> annoClass = cls.asSubclass(Annotation.class);
+
+                    // see if the annotation is supported by a checker
+                    if (isSupportedAnnotationClass(annoClass)) {
+                        return annoClass;
+                    }
+                }
+            }
+        } catch (ClassNotFoundException e) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: Try to use the minimal scopes for try/catch blocks. Here, is this only for the Class.forName in the first statement?

File: framework/src/org/checkerframework/framework/util/PurityAnnotatedTypeFactory.java (hunk 25 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/PurityAnnotatedTypeFactory.java]: Don't remove the class.
Try creating a PurityUnqualified like we did with the ReportUnqualified.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java (hunk 34 above)
Comment (jyluo) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: The set can be final and the call can be made at the constructor. I'll put in these changes and remove the boolean.

As for making the set immutable, I err on not doing so in the Loader, since the ATF still adds explicitly listed qualifiers and PolyAll to this set. The responsibility of making it immutable lies with the ATF at this moment. If the Loader returns an immutable set, then the ATF has to copy the elements in the set before adding the explicitly listed qualifiers and PolyAll.

What do you suggest?

Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java]: You should just document what sharing might exist. If the loadedAnnotations Set gets changed by the ATF, that change will be visible to other users of that Loader. So this interaction between the two objects needs some documentation.
